### PR TITLE
Add TCP protocol for GELF

### PR DESCRIPTION
**Pull request by edefaria**
Add TCP protocol support.

Based on #6,

### CODE REVIEW

**Comment by suyograo** [comment]
@edefaria thanks for this! can you please add some tests?

File: lib/logstash/inputs/gelf.rb [hunk 1]
```diff
@@ -33,6 +33,13 @@ class LogStash::Inputs::Gelf < LogStash::Inputs::Base
   # ports) may require root to use.
   config :port, :validate => :number, :default => 12201
 
+  # The GELF protocol (TCP or UDP).
+  config :protocol, :validate => :string, :default => "UDP"
+
+  # Whether to capture the hostname or numeric address of the incoming connection
+  # Defaults to hostname
+  config :use_numeric_client_addr, :validate => :boolean, :default => false
```
Comment (jordansissel) [inline lib/logstash/inputs/gelf.rb]: prefer not abbreviating words in config settings, use `address` instead of `addr`

File: lib/logstash/inputs/gelf.rb [hunk 2]
```diff
@@ -33,6 +33,13 @@ class LogStash::Inputs::Gelf < LogStash::Inputs::Base
   # ports) may require root to use.
   config :port, :validate => :number, :default => 12201
 
+  # The GELF protocol (TCP or UDP).
+  config :protocol, :validate => :string, :default => "UDP"
```
Comment (jordansissel) [inline lib/logstash/inputs/gelf.rb]: You can use `:validate => [ "UDP", "TCP" ]` to enforce a specific set of available values.

File: lib/logstash/inputs/gelf.rb [hunk 3]
```diff
@@ -80,20 +93,91 @@ def run(output_queue)
...
+              @jsonObj = JSON.parse(@data_in)
+            rescue => ex
+              @logger.warn("Gelf (tcp): failed to parse a message. Skipping: " + @data_in, :exception => ex, :backtrace => ex.backtrace)
+              next
+            end
+
+            begin  # Create event
+              event = LogStash::Event.new(@jsonObj)
+              event.remove("source_host")
+              event["source_host"] = @use_numeric_client_addr && client.addr(:numeric)[3] || client.addr(:hostname)[3]
```
Comment (jordansissel) [inline lib/logstash/inputs/gelf.rb]: `event[]=` is the old api won't be supported in Logstash 5.0. You'll want to use `event.set(field, value)`

**Comment by jordansissel** [comment]
Are there docs on the TCP protocol for GELF? I don't see any on the [gelf](http://docs.graylog.org/en/latest/pages/gelf.html) site.

**Comment by edefaria** [comment]
Gelf TCP is the same as UDP. The only thing is you need a delimit for your payload, by default on graylog/gelf is nul '\0'.
Also Graylog GELF TCP does not support compression, so most of the GELF library over TCP disable compression for TCP.

I will see to test and new api support.

**Comment by ddelnano** [comment]
Is there a timetable for if / when this will be merged??  Is there anything left to do, wouldn't mind helping if its needed.

**Comment by jordansissel** [comment]
@ddelnano It could help if some folks helped with any of the following:

* Test this in production/staging environments
* Evaluate it for edge cases
* Check that the new features are supported by new tests
* Check that the code is understandable, and ask questions in places where it is not.
* Check that the patch doesn't contain unrelated changes to the proposal
* For new features, do you agree that the setting names are the right names to choose? The right validation?
* For new features, are there docs that accompany it? Are the docs clear to you?

**Comment by ddelnano** [comment]
@jordansissel since July we have moved away from logstash, I appreciate the response tho.

**Comment by jordansissel** [comment]
This PR makes a number of changes that seem unrelated to adding TCP support:

* Several changelog, readme, etc changes.
* Many new specs that seem to cover things other than TCP support

I'm not sure what to make of this.

File: lib/logstash/inputs/gelf.rb [hunk 4]
```diff
@@ -33,6 +33,15 @@ class LogStash::Inputs::Gelf < LogStash::Inputs::Base
...
   config :port, :validate => :number, :default => 12201
 
+  # The GELF protocol (TCP or UDP).
+  config :protocol, :validate => [ "UDP", "TCP" ], :default => "UDP"
+
+  # Whether to capture the hostname or numeric address of the incoming connection
+  # in protocol tcp.
+  #
+  # Defaults to hostname
+  config :use_numeric_client_addr, :validate => :boolean, :default => false
```
Comment (jordansissel) [inline lib/logstash/inputs/gelf.rb]: I would prefer not abbreviating `address`.

I wonder if this could be a better name. Some questions for thinking about this:

* Do any other plugins have a similar option? 
* In what situations would a user want to set this to true?
* This setting seems, when `true`, to describe not-acting (meaning, to *not* do a dns lookup). Can we think of a positive-action name for this setting instead of a negative-action?

File: lib/logstash/inputs/gelf.rb [hunk 5]
```diff
@@ -33,6 +33,15 @@ class LogStash::Inputs::Gelf < LogStash::Inputs::Base
   # ports) may require root to use.
   config :port, :validate => :number, :default => 12201
 
+  # The GELF protocol (TCP or UDP).
+  config :protocol, :validate => [ "UDP", "TCP" ], :default => "UDP"
```
Comment (jordansissel) [inline lib/logstash/inputs/gelf.rb]: Could we support both protocols simultaneously instead of having this be a setting?

For example, the `syslog` input listens on both tcp and udp at the same time.

**Comment by olqs** [comment]
What's the status of this PR?
If it gets merged in the official plugin we don't have to build a custom one for us.

**Comment by acchen97** [comment]
Anyone who is using this plugin extensively want to step up and volunteer to help maintain this plugin?

https://discuss.elastic.co/t/logstash-plugins-community-maintainers/35953

**Comment by jordansissel** [comment]
This PR's been idle for some time. I'm going to close it because I don't know how to get clarity on the proposed change as compared to the patch provided.

We can still add TCP support for GELF input, but for whoever does, please include rspec tests, some indication of what product(s) you used to test it.

File: lib/logstash/inputs/gelf.rb (hunk 5 above)
Comment (azman0101) [reply lib/logstash/inputs/gelf.rb]: Mentioning that on doc would be great :)
